### PR TITLE
AJAX Comment 

### PR DESCRIPTION
**Pull request by iambibhas**

### CODE REVIEW

File: funnel/models/commentvote.py [hunk 1]
```diff
@@ -193,6 +193,7 @@ class Comment(UuidMixin, BaseMixin, db.Model):
                 'user',
                 'title',
                 'message',
+                'parent_id',
```
Comment (jace) [inline funnel/models/commentvote.py:196]: This column needs renaming. It should be `in_reply_to_id`. Also, public-facing id is `in_reply_to_uuid_b58`, a property.

Comment (jace) [reply funnel/models/commentvote.py:196]: It is unfortunate that the property will require loading the comment itself to retrieve its UUID. This is a good case for UUID primary keys.

File: funnel/templates/project_comments.html.jinja2 [hunk 2]
```diff
@@ -41,12 +41,52 @@
...
+            <div id="comments-wrapper">
+              <ul class="mui-list--unstyled" v-for="comment in comments">
+                <comment :comment="comment" :isuserparticipant="isuserparticipant"></comment>
+              </ul>
+            </div>
+          {% endraw %}
+          {% raw %}
+            <script type="text/x-template" id="comment-template">
+              <li class="comment">
+                <div id="c{{ comment.uuid_b58 }}">
```
Comment (jace) [inline funnel/templates/project_comments.html.jinja2:58]: Can we make this prefix `c.` or `c-` instead of `c`?

**Comment by vidya-ram** [comment]
- [ ] Add comments json endpoint to proposal page

- [ ] user details is missing in comments json

File: funnel/views/commentvote.py [hunk 3]
```diff
@@ -159,40 +160,88 @@ def view(self):
...
+                message=commentform.message.data,
+            )
+
+            self.obj.commentset.count = Commentset.count + 1
+            comment.voteset.vote(current_auth.user)  # Vote for your own comment
+            db.session.add(comment)
+            db.session.commit()
+            return {
+                'status': 'ok',
+                'message': _("Your reply has been posted"),
```
Comment (jace) [inline funnel/views/commentvote.py:182]: "Your comment has been posted"

File: funnel/templates/project_updates.html.jinja2 [hunk 4]
```diff
@@ -30,14 +30,14 @@
...
-            <a class="post-form" href="{{ project.url_for('new_update') }}" aria-label="Add new update" data-cy-btn="add-post">
-            <div class="user">
-              <div class="user__box">
-                <div class="user__box__gravatar user__box__gravatar--initials">{{ current_auth.user.fullname|initials }}</div>
-                <input type="text" placeholder="Post an update…" />
+            <a class="post-form-block" href="{{ project.url_for('new_update') }}" aria-label="Add new update" data-cy-btn="add-post">
+              <div class="user">
+                <div class="user__box">
+                  <div class="user__box__gravatar user__box__gravatar--initials">{{ current_auth.user.fullname|initials }}</div>
+                  <input type="text" placeholder="Post an update…" />
```
Comment (jace) [inline funnel/templates/project_updates.html.jinja2:37]: `{% trans %}` needed for placeholder.

File: funnel/templates/project_updates.html.jinja2 [hunk 5]
```diff
@@ -30,14 +30,14 @@
       <div class="grid" id="updates">
         <div class="grid__col-xs-12">
           {% if project.features.post_update() -%}
-            <a class="post-form" href="{{ project.url_for('new_update') }}" aria-label="Add new update" data-cy-btn="add-post">
-            <div class="user">
-              <div class="user__box">
-                <div class="user__box__gravatar user__box__gravatar--initials">{{ current_auth.user.fullname|initials }}</div>
-                <input type="text" placeholder="Post an update…" />
+            <a class="post-form-block" href="{{ project.url_for('new_update') }}" aria-label="Add new update" data-cy-btn="add-post">
```
Comment (jace) [inline funnel/templates/project_updates.html.jinja2:33]: aria-label needs trans tag

File: funnel/views/commentvote.py [hunk 6]
```diff
@@ -1,27 +1,50 @@
...
 
 ProposalComment = namedtuple('ProposalComment', ['proposal', 'comment'])
 
 
+@Comment.views('url')
+def comment_url(obj):
+    url = None
+    commentset_url = obj.commentset.views.url()
+    if commentset_url is not None:
+        url = commentset_url + '#c' + obj.uuid_b58
```
Comment (jace) [inline funnel/views/commentvote.py:32]: I'd like this to be `#c-` or `#c_`. We have to make this change in a few places around the code.

File: funnel/views/commentvote.py [hunk 7]
```diff
@@ -239,21 +319,42 @@ def delete_vote(self):
...
-                _("The comment has been reported as spam"), 'info',
-            )
-        else:
-            flash(
-                _("There was an issue reporting this comment. Please try again"),
-                'error',
-            )
-        return redirect(self.obj.commentset.views.url(), code=303)
+        if request.method == 'POST':
+            if csrf_form.validate_on_submit():
```
Comment (jace) [inline funnel/views/commentvote.py:327]: For a moment I was confused about why both conditions were being tested. Had to look a bit to notice there was POST + validation failure condition below. Suggestion: use `csrf_form.validate()` here to make it not look accidental.

File: funnel/templates/macros.html.jinja2 [hunk 8]
```diff
@@ -767,6 +767,80 @@
...
+        </div>
+      </a>
+      <a class="mui-btn mui-btn--small mui-btn--raised mui-btn--primary" :href="loginUrl" data-cy="login-btn" v-if="!isuserloggedin">Login to leave a comment</a>
+    </div>
+    <script type="text/x-template" id="comment-template">
+      <li class="comment">
+        <div :id="'#c-' + comment.uuid_b58">
+          <div class="comment__header">
+            <button class="mui-btn mui-btn--nostyle" aria-label="Hide comment" aria-expanded="true" v-if="!hide" @click="collapse(true)">
+              <svg class="fa5-icon fa5-icon--subhead test fa5--align-baseline" aria-hidden="true" role="img"><use v-bind:xlink:href="svgIconUrl + '#caret-circle-up'"></use></svg>
```
[the remark this answers is not in the document]
Comment (vidya-ram) [reply funnel/templates/macros.html.jinja2:804]: Sure, opening an issue for this.

[the remark this answers is not in the document]
Comment (vidya-ram) [reply funnel/templates/macros.html.jinja2:804]: https://github.com/hasgeek/funnel/issues/822

**Comment by vidya-ram** [comment]
Comments json endpoint for proposal needs fixing

**Pull request merged**